src/neoxp/Node/NodeUtility.cs

```diff
@@ -208,6 +211,197 @@ public static void SignOracleResponseTransaction(ProtocolSettings settings, Expr
+
+        public static int PersistContract(NeoSystem neoSystem, ContractState state,
+            IReadOnlyList<(string key, string value)> storagePairs, ContractCommand.OverwriteForce force)
+        {
+            var (overwriteContract, overwriteStorage) = force switch 
+            {
+                ContractCommand.OverwriteForce.All => (true, true),
+                ContractCommand.OverwriteForce.ContractOnly => (true, false),
+                ContractCommand.OverwriteForce.None => (false, false),
+                ContractCommand.OverwriteForce.StorageOnly => (false, false),
+                _ => throw new NotSupportedException($"Invalid OverwriteForce value {force}"),
+            };
+
+            using var snapshot = neoSystem.GetSnapshot();
+
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_Contract).Add(state.Hash);
+            var localContract = snapshot.GetAndChange(key)?.GetInteroperable<ContractState>();
+            if (localContract is null)
+            {
+                // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+                // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+                // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+                // over keys that shouldn't exist for one of the contracts.
+                if (state.Id <= LastUsedContractId(snapshot))
+                {
+                    state.Id = GetNextAvailableId(snapshot);
+                }
+                else
+                {
+                    // Update available id such that a regular contract deploy will use the right next id;
+                    SetNextAvailableContractId(snapshot, state.Id + 1);
+                }
+
+                snapshot.Add(key, new StorageItem(state));
+            }
+            else
+            {
+                if (overwriteContract)
+                {
+                    // Note: a ManagementContract.Update() will not change the contract hash. Not even if the NEF changed.
+                    localContract.Nef = state.Nef;
+                    localContract.Manifest = state.Manifest;
+                    localContract.UpdateCounter = state.UpdateCounter;
+                }
+                else
+                {
+                    // check to make sure the existing contract matches the downloaded contract
+                    // except for id, which is allowed to be different from downloaded contract
+                    if (!localContract.Hash.Equals(state.Hash)
+                        || localContract.UpdateCounter != state.UpdateCounter
+                        || !localContract.Nef.ToArray().SequenceEqual(state.Nef.ToArray())
+                        || !localContract.Manifest.ToJson().ToByteArray(false).SequenceEqual(state.Manifest.ToJson().ToByteArray(false)))
```

Comment:
This part feels strange to have. If we reach this point then the following conditions hold
* We found a contract locally with the same `hash`  as `state.hash` (the remote contract) at L313. Therefore 
   ```C#
   !localContract.Hash.Equals(state.Hash)
   ```
   is always `false`
* the `force` argument is `StorageOnly` || `NONE`
   * if `force` = `StorageOnly` we do nothing because the logic between lines 356-399 deals with storage
   * if `force` = `NONE` then we're not supposed to change anything and should always fail. 

I think the only logic that should be here is the detailed explanation what part of the contract already exists (as I thought was requested here https://github.com/neo-project/neo-express/pull/209#discussion_r845574402) but those changes are now gone and we actually get a generic error instead.